Fix handling of `allow_bottom_tparams` in `isambiguous`

timholy opened the pull request:
With master, we have a lot of ambiguities, but a subset of these involve method signatures that are ambiguous only because of the presence of `Union{}` as a type-parameter in the intersection:
```julia
julia> foo(x::Complex{<:Integer}) = 1
foo (generic function with 1 method)

julia> foo(x::Complex{<:Rational}) = 2
foo (generic function with 2 methods)

julia> m1, m2 = collect(methods(foo))
2-element Array{Any,1}:
 foo(x::Complex{#s1} where #s1<:Integer) in Main at REPL[3]:1 
 foo(x::Complex{#s1} where #s1<:Rational) in Main at REPL[4]:1

julia> typeintersect(m1.sig, m2.sig)
Tuple{#foo,Complex{Union{}}}

julia> Base.isambiguous(m1, m2)
true
```
This isn't the kind of intersection we should usually be worried about, so in e101000094794b8a14502c294ab75248cbba0be8 a strategy was developed to exclude these kinds of intersections. However, I found that it doesn't correctly catch all cases, including the one above:
```
julia> Base.isambiguous(m1, m2, false)
true
```
With this PR, I believe I've fixed the detection of `Union{}`. It has a pretty dramatic effect on our ambiguity list.

Master:
```julia
julia> length(Base.Test.detect_ambiguities(Base,Core,allow_bottom=true))
Skipping Base.<|
Skipping Base.Parallel
194

julia> length(Base.Test.detect_ambiguities(Base,Core,allow_bottom=false))
Skipping Base.<|
Skipping Base.Parallel
111
```

This PR:
```julia
julia> length(Base.Test.detect_ambiguities(Base,Core,allow_bottom=true))
Skipping Base.<|
Skipping Base.Parallel
194

julia> length(Base.Test.detect_ambiguities(Base,Core,allow_bottom=false))
Skipping Base.<|
Skipping Base.Parallel
11
```

11 seems rather more manageable (the "missing" digit is *not* a copy/paste error). Suddenly the steps that @Sacha0 has been taking (#20970, #20971) don't seem so small anymore: those should fix 3 of the 11.  Of the rest, it's just 4 functions: `cat` (4 ambiguous pairs), `checkbounds_indices` (2 ambiguous pairs), `colon` (1 ambiguous pair), and a constructor method for `Tuple` (1 ambiguous pair).

I do find myself wishing we could change the keyword argument to consistently have the same name, and I find `accept_bottom_tparams` clearer than `allow_bottom_tparams` (and perhaps flipping the sense, `exclude_bottom_tparams`, to be the clearest of all). But at this point, I don't think we can do that for 0.6. I did add some docstrings.

Fixes #20246

tkelman commented on base/reflection.jl:
````diff
@@ -941,13 +941,44 @@ function method_exists(f::ANY, t::ANY)
+
+For parametric types, `allow_bottom_tparams` controls whether
+`Union{}` is considered a valid intersection of type parameters. For example:
+```julia
````
looks like this could be a doctest with minor changes

Sacha0 reviewed (approved):
Insofar as I am qualified to review (not far), this looks great!

Sacha0 commented:
(#20970, #20971, #20986, #20988, #20989, and #20990 should together resolve all relevant ambiguities. Best!)

timholy committed:
Fix handling of `allow_bottom_tparams` in `isambiguous`

The former version failed to detect that certain type parameters were Bottom (for example, in the case in the added docstring for `isambiguous`).

timholy commented:
@Sacha0, when your pile of PRs merges, please do us the honor of turning that test back on. Thanks for jump-starting the effort to resolve these ambiguities.

Sacha0 commented:
> `exclude_bottom_tparams`, to be the clearest of all

+1

tkelman commented on base/reflection.jl:
```diff
@@ -941,13 +941,41 @@ function method_exists(f::ANY, t::ANY)
 end
 
+"""
+    isambiguous(m1, m2, [allow_bottom_tparams=true]) -> Bool
```
unrelated aside, but I think we should gradually aim to get away from using square brackets to denote optional arguments and uniformly use the julia syntax convention for them in docstrings
